File: Algorithms/p160_Intersection_of_Two_Linked_Lists/p160_Intersection_of_Two_Linked_Lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        int len_a = 0;
        for(ListNode* p=headA;p;p=p->next) ++len_a;
        int len_b = 0;
        for(ListNode* p=headB;p;p=p->next) ++len_b;
        ListNode* pa = headA;
        ListNode* pb = headB;
        if(len_a>len_b){
            for(int i=len_a-len_b;i>0;--i) pa = pa->next;
        }else if(len_b>len_a){
            for(int i=len_b-len_a;i>0;--i) pb = pb->next;
        }
        while(pa!=pb){
            pa = pa->next;
            pb = pb->next;
        }
        return pa;
    }
};

File: Algorithms/p160_Intersection_of_Two_Linked_Lists/p160_Intersection_of_Two_Linked_Lists.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def getIntersectionNode(self, headA, headB):
        """
        :type head1, head1: ListNode
        :rtype: ListNode
        """
        len_a = 0
        p = headA
        while p:
            len_a += 1
            p = p.next
        len_b = 0
        p = headB
        while p:
            len_b += 1
            p = p.next
        pa = headA
        pb = headB
        if len_a>len_b:
            for i in range(len_a-len_b):
                pa = pa.next
        elif len_b>len_a:
            for i in range(len_b-len_a):
                pb = pb.next
        while pa!=pb:
            pa = pa.next
            pb = pb.next
        return pa
